core/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -579,23 +573,54 @@ private Updates changes(Snapshot snapshot) {
-                    if (updates.startedShards.isEmpty() == false || updates.failedShards.isEmpty() == false) {
+
+                    if (updates.isEmpty() == false) {
                         ImmutableOpenMap.Builder<ShardId, ShardRestoreStatus> shardsBuilder = ImmutableOpenMap.builder(entry.shards());
-                        for (Map.Entry<ShardId, ShardRestoreStatus> startedShardEntry : updates.startedShards.entrySet()) {
-                            shardsBuilder.put(startedShardEntry.getKey(), startedShardEntry.getValue());
+                        for (Map.Entry<ShardId, ShardRestoreStatus> startedShard : updates.startedShards.entrySet()) {
+                            shardsBuilder.put(startedShard.getKey(), startedShard.getValue());
                         }
-                        for (Map.Entry<ShardId, ShardRestoreStatus> failedShardEntry : updates.failedShards.entrySet()) {
-                            shardsBuilder.put(failedShardEntry.getKey(), failedShardEntry.getValue());
+                        for (Map.Entry<ShardId, ShardRestoreStatus> failedShard : updates.failedShards.entrySet()) {
+                            shardsBuilder.put(failedShard.getKey(), failedShard.getValue());
                         }
+                        for (Map.Entry<ShardId, UnassignedInfo> unassignedShard : updates.unassignedShards.entrySet()) {
+                            ShardId shardId = unassignedShard.getKey();
+                            UnassignedInfo unassignedInfo = unassignedShard.getValue();
+                            String currentNodeId = routingTables.shardRoutingTable(shardId).primaryShard().currentNodeId();
+
+                            if (unassignedInfo.getFailure() != null && Lucene.isCorruptionException(unassignedInfo.getFailure().getCause())) {
+                                // mark restore entry for this shard as failed when it's due to a file corruption. There is no need wait on retries
+                                // to restore this shard on another node if the snapshot files are corrupt. In case where a node just left or crashed,
+                                // however, we only want to acknowledge the restore operation once it has been successfully restored on another node.
+                                shardsBuilder.put(shardId, new ShardRestoreStatus(currentNodeId, RestoreInProgress.State.FAILURE,
+                                    unassignedInfo.getFailure().getCause().getMessage()));
+                            } else {
+                                // check if the maximum number of attempts to restore the shard has been reached. If so, we can fail
+                                // the restore and leave the shards unassigned.
+                                IndexMetaData indexMetaData = metaData.getIndexSafe(unassignedShard.getKey().getIndex());
+                                int maxRetry = MaxRetryAllocationDecider.SETTING_ALLOCATION_MAX_RETRY.get(indexMetaData.getSettings());
```

Comment:
I think the solution needs to be more generic than depending on the settings of specific allocation deciders. I think we can use `unassignedInfo.getLastAllocationStatus` for that and check if it is `DECIDERS_NO`.

Reply:
That's a good suggestion as I suppose that a restore can also be stuck because the deciders cannot assign the shard (no enough space on disk, awareness rules forbid allocation etc). I also like it to be more generic.

I think I can give it a try by reverted portion of code and override `unassignedInfoUpdated()`... I'll push something if it works.